server/src/main/java/org/opensearch/index/store/CompositeDirectory.java

```diff
@@ -0,0 +1,292 @@
+        } else {
+            fileLength = remoteDirectory.fileLength(name);
+            logger.trace("fileLength from Remote {}", fileLength);
+        }
+        return fileLength;
+    }
+
+    /**
+     * Creates a new, empty file in the directory and returns an {@link IndexOutput} instance for
+     * appending data to this file.
+     * @param name the name of the file to create.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        logger.trace("createOutput() called {}", name);
+        // The CloseableFilterIndexOutput will ensure that the file is added to FileCache once write is completed on this file
+        return new CloseableFilterIndexOutput(localDirectory.createOutput(name, context), name, this::cacheFile);
+    }
+
+    /**
+     * Ensures that any writes to these files are moved to stable storage (made durable).
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void sync(Collection<String> names) throws IOException {
+        logger.trace("sync() called {}", names);
+        Collection<String> remoteFiles = Arrays.asList(getRemoteFiles());
+        Collection<String> filesToSync = names.stream().filter(name -> remoteFiles.contains(name) == false).collect(Collectors.toList());
+        logger.trace("Synced files : {}", filesToSync);
+        localDirectory.sync(filesToSync);
+    }
+
+    /**
+     * Renames {@code source} file to {@code dest} file where {@code dest} must not already exist in
+     * the directory.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        logger.trace("rename() called {}, {}", source, dest);
+        localDirectory.rename(source, dest);
+        fileCache.remove(localDirectory.getDirectory().resolve(source));
+        cacheFile(dest);
+        fileCache.decRef(localDirectory.getDirectory().resolve(dest));
+    }
+
+    /**
+     * Opens a stream for reading an existing file.
+     * Check whether the file is present locally or in remote and return the IndexInput accordingly
+     * @param name the name of an existing file.
+     * @throws IOException in case of I/O error
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        logger.trace("openInput() called {}", name);
+        // We aren't tracking temporary files (created via createTempOutput) currently in FileCache as these are created and then deleted
+        // within a very short span of time
+        // We will be reading them directory from the local directory
+        if (isTempFile(name)) {
+            return localDirectory.openInput(name, context);
+        }
+        // Return directly from the FileCache (via TransferManager) if complete file is present
+        Path key = localDirectory.getDirectory().resolve(name);
```

Comment:
can there be a case where key is returned as null? if yes, we should add a null check

Reply:
Key won't be null if local directory is not null. We already have checks for non null values of local directory in the Composite Directory constructor